Fix BaseFileCacheModule#keys to respect prefix

Urth opened the pull request:
##### SUMMARY
Previously BaseFileCacheModule#keys would return keys with the cache prefix. These keys are impossible to retrieve from the cache without removing the prefix or using the cache without a prefix. Now it removes the prefix from the key and only returns keys that share the same prefix as the cache.

##### ISSUE TYPE
- Bugfix Pull Request
- Test Pull Request

##### COMPONENT NAME
cache

##### ADDITIONAL INFORMATION
New tests without the fix for all BaseFileCacheModule based caches.

```
$ ansible-test units --docker -v --python 3.9 test/units/plugins/cache/test_cache.py
=================================== FAILURES ===================================
______________________ TestJsonFileCachePrefix.test_flush ______________________
[gw1] linux -- Python 3.9.0 /tmp/python-yclxd6af-ansible/python
self = <units.plugins.cache.test_cache.TestJsonFileCachePrefix testMethod=test_flush>

    def test_flush(self):
        # Fake that the cache already has some data in it but the adjudicator
        # hasn't loaded it in.
        self.cache._plugin.set('monkey', 'animal')
        self.cache._plugin.set('wolf', 'animal')
        self.cache._plugin.set('another wolf', 'another animal')
    
        # The adjudicator does't know about the new entries
        assert len(self.cache.keys()) == 2
        # But the cache itself does
        assert len(self.cache._plugin.keys()) == 3
    
        # If we call flush, both the adjudicator and the cache should flush
        self.cache.flush()
        assert len(self.cache.keys()) == 0
>       assert len(self.cache._plugin.keys()) == 0
E       AssertionError: assert 3 == 0
E         +3
E         -0

test/units/plugins/cache/test_cache.py:108: AssertionError
```

Urth committed:
Fix BaseFileCacheModule#keys to respect prefix

Change:
- Previously BaseFileCacheModule#keys would return keys with the cache
  prefix. These keys are impossible to retrieve from the cache without
  removing the prefix or using the cache without a prefix.
  Now it removes the prefix from the key and only returns keys that
  share the same prefix as the cache.

Test Plan:
- New unit tests

s-hertel commented on test/units/plugins/cache/test_cache.py:
```diff
@@ -86,18 +98,36 @@ def test_flush(self):
+        shutil.rmtree(self.cache_dir)
+
+
+class TestJsonFileCachePrefix(TestJsonFileCache):
```
Are you planning on adding a test to make sure keys() only retrieves those with the correct prefix? That would be a great addition.

Urth replied:
Sure I'll try to get that added before the holidays

s-hertel reviewed (approved):
Thanks for fixing this. It looks like the _cache dictionary is expected to contain the keys without the prefix, and setting/getting/deleting updates the key with the prefix to access the actual file.

Urth committed:
Add explicit BaseFileCacheModule#keys test

Test that files that do not match the cache prefix are ignored.
Test that the prefix is removed from the cache key.

s-hertel commented on lib/ansible/plugins/cache/__init__.py:
```diff
@@ -207,10 +207,20 @@ def has_expired(self, key):
+        # When using a prefix we must remove it from the key name before
+        # checking the expiry and returning it to the caller. Keys that do not
+        # share the same prefix cannot be fetched from the cache.
+        prefix = self.get_option('_prefix')
```
Sorry for missing your update. The tests look good, thanks for including those. I just noticed that the prefix can be None, which will fail with `len`.
```suggestion
        prefix = self.get_option('_prefix') or ''
```

bcoca replied:
instead of `or ''` just set the default to `''`, you should never need conditionals on `get_option`, otherwise the docs wont really reflect the real options

bcoca replied:
also might need to add aliases .. _prefix should be prefix since it is not hidden/internal/unsettable option (my mistake from initial conversion of the plugins, kept those names)

bcoca replied:
another random thought ... we need to specify 'required common options' per plugin type, in this case 'prefix' is required for cache plugins.

s-hertel replied:
No, both inventory and fact cache have non-null defaults for prefix already. It occured to me someone might do:
```
# ansible.cfg
[inventory]
cache_plugin=jsonfile
cache_connection=/tmp/mycache
cache_prefix=
```
And then the only other place we care about prefix handles this gracefully already:
```
    def _get_cache_file_name(self, key):
        prefix = self.get_option('_prefix')
        if prefix:
            cachefile = "%s/%s%s" % (self._cache_dir, prefix, key)
        else:
            cachefile = "%s/%s" % (self._cache_dir, key)
        return cachefile
```
Good point about aliases, I can update as a follow-up PR.

s-hertel replied:
It looks like it's only an empty string though in that case though, so I rescind that

s-hertel commented:
Thanks @Urth!
